[IV-23-24] Objetivo 9

javigp2002 opened the pull request:
✨ Objetivo 9 y fin de IV incoming ✨

javigp2002 committed:
:sparkles: Init: crea bd temporal con kv closes #75

Se va a crear una base de datos con KV donde podremos almacenar tuplas clave valor con los equipos y los jugadores de los mismos para poder devolverlos, crear, modificar o eliminar en una bd temporal

javigp2002 committed:
:rocket: Add representacion: interfaces para la representacion de datos. closes #76

Representar los datos de una manera çoptima es fundamental para poder trabajar de cara a la lógica de negocio y, por ello, se crean las 3 partes fundamentales y se recoge el calendario en una de ellas

javigp2002 committed:
:rocket: funciones de la base de datos, closes #77

Estas son las funciones que van a necesitarse de cara al REST API cuando se pida información de los datos 'almacenados', se trabaja y se devuelve en JSON

javigp2002 committed:
:white_check_mark: Add test bd, closes #78

En estos tests se comprobará la integridad de la base de datos así como las interfaces creadas que se asemejan a las empleadas en la lógica de negocio para que en un futuro se mantengan esos modelos o se cambie completamente todo

javigp2002 committed:
:fire: Update: funciones de la API, closes #79

Con estas funciones y con la base de datos ahora somos capaces de devolver en todo momento el ok o los datos necesarios a sacar al usuario que lo necesite según las HU

javigp2002 committed:
:bug: Delete: función sin lógica de negocio, closes #80

Esta función carecía de sentido en torno a la lógica de negocio

javigp2002 committed:
:white_check_mark: Add tests api, closes #81

La comprobación principal es que al enviar la información necesaria esta es comprobada y devuelta con éxito, independientemente de lo que haga por dentro

javigp2002 committed:
:rocket: Add info de paquetes, closes #82

Actualización del deno.lock para los paquetes introducidos en las últimas versiones

javigp2002 committed:
:zap: Introducido para pasar el objetivo, closes #83

No lo he necesitado pero ha sido necesario para poder utilizarlo

javigp2002 committed:
:zap: Add configuración para realizar los tests, closes #74, closes #73

Con esto se pone fin a los primeros cambios sobre el objetivo y se cambia también en el dockerfile puesto que este llama a 'deno task test' que ya lo tiene almacenado el como realizarlo

javigp2002 committed:
:zap: Improving: separación de lógica API, closes #85

Se ha creado un controlador de la api en la cual residirá la lógica de negocio

javigp2002 committed:
:white_check_mark: Add testing for controller, closes #86

Confirmarmos que  se siga la lógica de negocio a traves de un mocking (nos da igual si devuelve bien la bd puesto que tiene su propio test)

javigp2002 committed:
:zap: clase de Base de datos, closes #87

La base de datos tiene todas las funcionalidades necesarias para poder realizar cualquier busqueda a través de la Api

javigp2002 committed:
:white_check_mark: Add testing for API, closes #88

Creación de test para la API donde comprobaremos tests de integración

javigp2002 committed:
:white_check_mark: Add testing for bd, closes #89

Testeamos que la base de datos esté aplicando la lógica de negocio necesaria cada vex que se introduce, cambia o crea (como kv está en fase de pruebas se tienen q eliminar los que se crean puesto que no se puede reiniciar la bbdd)

javigp2002 committed:
:zap: Improve API con controlador, closes #90

Ahora la API no se encarga de la lógica trás las peticiones sino simplemente de manejar las mismas

javigp2002 committed:
Add: dependencias, closes #82

Añadidas las nuevas dependencias

javigp2002 committed:
:bug: Fixing bug test_Bd

Ha saltado un error que no había saltado en local

javigp2002 committed:
:bug: Fixing again bug test_Bd

Error en la creación de un jugador tipo

danieeeld2 reviewed (commented):
No lo he mirado tampoco con mucha profundidad, porque estoy liado con los exámenes, pero parece que esta bastante correcto. Le he prestado más atención a la parte de la API y parece que está correcta. Además, se ve claramente que endpoint aplica la lógica de negocio. Como comentario, para la parte de los tests, para que fueran más legibles, podría ser buena idea que leyera los datos que usas para testar de un fichero a parte, en vez de declararlos todos ahí como constantes :+1: :tada:

javigp2002 commented:
Gracias por tu comentario, sí sería más óptimo y legible como hacia en los primeros tests, pero en este periodo que nos encontramos el tiempo para optimizar estas cosillas escasea aunque sería ideal. Muchas gracias @danieeeld2 por tu revisión, @JJ listo para revisión.

JJ commented on api/api_controller.ts:
```diff
@@ -0,0 +1,58 @@
+        return res;
+    }
+
+    async deleteEquipo(nombreEquipo: string) {
```
¿Qué historia de usuario pide borrar un equipo?

javigp2002 replied:
Cierto, Resuelto en [2383612](https://github.com/javigp2002/LazyFood/pull/84/commits/23836120b56c54b73ebf2e33b20ba8e0fa4ef9cc)

JJ commented on model/bd.ts:
```diff
@@ -0,0 +1,193 @@
+
+     async postJugador(body: any){
+        if (await this.existeJugador(body.nombre)){
+            return {ok: false};
```
Primero, usar `POST` para poner toda la información del jugador es cuestionable. ¿Por qué no usas PUT? Segundo, tienes que usar siempre status de HTTP, no mensajes a caso hecho.

javigp2002 replied:
He resuelto estos dos comentarios en [238ac08](https://github.com/javigp2002/LazyFood/pull/84/commits/238ac08d7ead6a396c5a2987ee8cc6654ffae77c), [587c8f4](https://github.com/javigp2002/LazyFood/pull/84/commits/587c8f42842867a96625f05465e8b6fbcca714ee) y [8f6b96f](https://github.com/javigp2002/LazyFood/pull/84/commits/8f6b96f3355d6c05ab24366e3a377c13c2efeabf). Cierto es que debía devolver el status pero no veo que se haga desde la DB, como he creado un API controller será este el que la maneje y devuelva estos datos al API.

JJ replied:
Si devuelve los datos al API, debe también devolverlos en un formato razonable. Tienes que ver qué es lo que necesita esta función: que cuando termine, el jugador esté creado. Si está creado *al principio*, lo más razonable es simplemente que retorne sin hacer nada, ¿no te parece? Si no, ¿qué va a hacer con esta información quien la reciba?

JJ commented on model/bd.ts:
```diff
@@ -0,0 +1,193 @@
+        return await this.kv.set([nombreJugadoresKv, body.nombre], body);
+    }
+
+     async postEquipo(body: any){
```
Igual que arriba. Si ya sabes toda la información del equipo, conviene que uses `PUT`. Se usa `PUT` si se conoce el URI, POST si no. 
En tu caso, el URI podía ser el nombre, que es único, ¿no?

javigp2002 replied:
Resuelto en https://github.com/javigp2002/LazyFood/pull/84#discussion_r1448176613

JJ commented on model/bd.ts:
```diff
@@ -0,0 +1,193 @@
+
+
+        const jugador = res.value as IJugador;
+        for (const key in body) {
```
Uf, usando el body para cosas que deberían estar en el URI, y usando PUT para modificar parcialmente un URI... No lo acabo de ver. El verbo que se debería usar para esto es `PATCH`. En todo caso, `PUT` es idempotente, así que no puedes devolver el mensaje de arriba si ya existe.

javigp2002 replied:
Es que estaban mal puestos los nombres de las variables, pero sí, los nombres son únicos. [238ac08](https://github.com/javigp2002/LazyFood/pull/84/commits/238ac08d7ead6a396c5a2987ee8cc6654ffae77c)

JJ reviewed (changes requested):
Me temo que va a hacer falta otra vuelta, sobre todo para que trates de aplicar mejor las convenciones de los APIs REST... Si tienes tiempo de hacerle los cambios que ha comentado @danieeeld2 , mejor.

javigp2002 committed:
:zap: Improve entrada de datos, closes #91

Según la recomendación de @danieeeld2 es mejor declarar en un archivo aparte los datos testeables para su legibilidad

javigp2002 committed:
:bug: Delete función sin relación a HU, closes #80

Ciertamente el delete no tiene ningún sentido para las HUs requeridas

javigp2002 committed:
:zap: Rename funciones de bd, closes #92

Los nombres que se habían propuesto estaban demasiado ligados a los de la API REST lo que puede llevar a un error en la lectura de los mismos

javigp2002 committed:
:rocket: Check status on api

Ahora también se devuelve tanto en el post como en el put el status del http

javigp2002 committed:
:white_check_mark: Add testing for API

Creamos los tests para confirmar que los nuevos cambios estan testeables

javigp2002 committed:
:bug: Delete unusued function, closes #91

Se ha eliminado la función que se utilizaba previa a la separación de datos

javigp2002 committed:
:bug: Fixing error en bd_test

Han fallado los tests

javigp2002 committed:
:bug: Seguimos fixing error en bd_test

Han fallado los tests

javigp2002 committed:
:bug: Seguimos fixing error en bd_test x3

Han fallado los tests

javigp2002 commented:
@JJ  Listo para revisión! Esta siendo dura la cuesta de enero... También he intentado resolver lo que me comentaba Daniel

JJ commented on model/bd.ts:
```diff
@@ -0,0 +1,193 @@
+
+     async createEquipo(data: any){
+        if (await this.existeEquipo(data.nombre)){
+            return {ok: false};
```
Igual que arriba. No puedes crear mensajes de error específicos como valor de retorno de una función. Si realmente piensas que es un error, lanza una excepción. En este caso, como en el anterior, sinceramente no creo que lo sea. Si ya está creado, que vuelva y listo. Pero es que ni siquiera: la operación de introducir en Kv no es "pesada", así que si hay que repetir la operación debería devolver el mismo resultado.

JJ replied:
Si *realmente* quieres devolver algo en caso de que ya esté creado (que ya te digo que es innecesario) la forma correcta sería, primero, que especifiques un tipo de vuelta con unos estados bien definidos o segundo que simplemente devuelvas un Boolean.
En todo caso, con un nombre como `create` no se espera que se devuelva nada, y tal como está definida la función quien la use no espera que se devuelva nada, así que lo más probable (y realmente posible por como está definida) es que lo que se devuelva simplemente se ignore. Que, una vez más, tiene sentido, porque si la intención era crear, ya está creado...

En todo caso, me gustaría que lo reconsideraras. Por el SRP, la responsabilidad de esta función es crear, no crear-comprobando-previamente-si-está-creado. Eso son *dos* responsabilidades, que deberían ir en *dos* funciones diferentes.

javigp2002 replied:
Ah vale ahora si entiendo lo que me comenta. Me pongo con ello

javigp2002 replied:
Resuelto en [3175851](https://github.com/javigp2002/LazyFood/pull/84/commits/317585162096d2ab8dcbabbbae4d28fc9442b971)

JJ reviewed (changes requested):
Has currado un montón en esto, y se nota. El único problema grave es el `listen` y el meter datos fijos en un módulo de desarrollo (en vez de ponerlos en test), pero hay otra serie de pequeñas cuestions, desde temas de *naming*, organización de los directorios, los dos APIs que tienes (con la BD y con el API REST), uso de mocks donde parece que no hacen falta... 
En fin, sé que la cuesta es dura, pero no me cabe duda de que le darás una vuelta y saldrá bien, todavía tienes una semanita (casi)

javigp2002 committed:
:ambulance: Delete unusued listen en api, closes https://github.com/javigp2002/LazyFood/pull/84#discussion_r1448491303

Es cierto que no aportaba nada de cara al módulo ni a los tests, por tanto, solo puede acarrear problemas

javigp2002 committed:
:recycle: Refactor api module, closes https://github.com/javigp2002/LazyFood/pull/84#discussion_r1448502027

La parte del API se considera del fuente y, por ello, debe estar en la misma carpeta

javigp2002 committed:
:recycle: Refactor api module, closes https://github.com/javigp2002/LazyFood/pull/84#discussion_r1448502027

La parte del API se considera del fuente y, por ello, debe estar en la misma carpeta

javigp2002 committed:
:recycle: Refactor nombre devuelto, closes https://github.com/javigp2002/LazyFood/pull/84#discussion_r1448505731

Es cierto que da lugar a confusión haber creado un mock que no tiene sentido literal lo que realiza

javigp2002 committed:
:rocket: Swap: valores de test en tests, closes https://github.com/javigp2002/LazyFood/pull/84#discussion_r1448515957

En este caso se va a crear los calendarios donde se necesitan y no en los sitios de producción, falta el calendario real puesto que ese sigue siendo de prueba

javigp2002 commented on api/api.ts:
```diff
@@ -1,65 +1,93 @@
-
 app.use(router.routes());
 app.use(router.allowedMethods());
+app.listen({ port: 8000 });
```
Cierto, no hacía falta [dc06a43](https://github.com/javigp2002/LazyFood/pull/84/commits/dc06a43c435a185d4bc0300ce2596fa17eed3ef8)

javigp2002 commented on test/api_test.ts:
```diff
@@ -0,0 +1,100 @@
+import { describe,it,beforeAll, beforeEach, afterAll } from "https://deno.land/std@0.204.0/testing/bdd.ts";
+import { assert, assertInstanceOf, assertEquals } from "https://deno.land/std@0.205.0/assert/mod.ts";
+import { superoak } from "https://deno.land/x/superoak/mod.ts";
+import { app } from "../api/api.ts";
```
[dc06a43](https://github.com/javigp2002/LazyFood/pull/84/commits/dc06a43c435a185d4bc0300ce2596fa17eed3ef8)

javigp2002 commented on test/api_controller_test.ts:
```diff
@@ -0,0 +1,61 @@
+import { ApiController } from "../api/api_controller.ts";
+import { MyDb, crearCalendario } from "../model/bd.ts";
+import  mockito  from "npm:ts-mockito";
+import { Jugador } from "../src/jugador.ts";
```
Vale, gracias. Resuelto en [387c3c5](https://github.com/javigp2002/LazyFood/pull/84/commits/387c3c552512d176fbf51304d48746e7cc29dd62), [b2cb19c](https://github.com/javigp2002/LazyFood/pull/84/commits/b2cb19cfba79c1652b4b66b6a425a7be755cf5b0)

javigp2002 commented on test/api_controller_test.ts:
```diff
@@ -0,0 +1,61 @@
+    let unJugador: Jugador;
+    
+    beforeAll(() => {
+        const mockedJugador = mockito.mock(Jugador);
```
LA verdad que solo he hecho un mock para probar como funcionan y porque me parecen muy útiles a la hora de hacer test unitarios. Con esto, según me informé, consigo que solo se testee la parte del código que necesito en ese momento puesto que los atributos o, el propio jugador en este caso, no afectan al test en sí. 

Es decir, me parecía más intuitivo y quería probarlos y, debido a que no era parte fundamental del objetivo, quería emplearlos de alguna manera. Pero sí, podría utilizar un `Jugador`

javigp2002 commented on test/api_controller_test.ts:
```diff
@@ -0,0 +1,61 @@
+
+    it ("M5.3 - Testing Get (Jugador Óptimo vacio)", async () => {
+        //when
+        const result = await underTest.getOptimo("   ");
```
Este, como he explicado antes, es simplemente por probar como funcionaban el mock 😄

javigp2002 commented on model/bd.ts:
```diff
@@ -0,0 +1,193 @@
+const nombreEquiposKv = "equipos";
+const nombreJugadoresKv = "jugadores";
+const date: Date = new Date("2021-01-01");
+const barcelona = new EquipoReal("Barcelona", 3);
```
Claramente no 👇 , [c8c8596](https://github.com/javigp2002/LazyFood/pull/84/commits/c8c859614d1437421b30855527701ac598438262)

javigp2002 commented:
@jj Listo para la ¿última? revisión 😄. He cambiado los problemas graves para pasar el objetivo y le he explicado el porque de los demás debido a que veo este último objetivo como una oportunidad para poder utilizar nuevas formas para los test unitarios.

javigp2002 commented on model/bd.ts:
```diff
@@ -0,0 +1,193 @@
+
+const nombreEquiposKv = "equipos";
+const nombreJugadoresKv = "jugadores";
+const date: Date = new Date("2021-01-01");
```
No la verdad, 💩 [c8c8596](https://github.com/javigp2002/LazyFood/pull/84/commits/c8c859614d1437421b30855527701ac598438262)

JJ commented on model/bd.ts:
```diff
@@ -0,0 +1,180 @@
+
+        const team = new Equipo(iTeam.nombre, jugadores, this.calendario);
+
+        if (date == undefined) date = new Date();
```
Vamos a ver, Javi, no hagas trampas en el solitario. ¿Qué pasa si quieres obtener el jugador óptimo de hace tres jornadas, porque se te pasó hacerlo?
Igual esto corresponde correctamente a la historia de usuario, pero si es así, una vez más, la denominación es importante: `getJugadores` primero transmite que quieres recuperar algo que ya existe, y segundo no da la idea de que vas a optimizar los jugadores para la fecha actual.

javigp2002 replied:
Pero con la HU y el la propuesta no tiene sentido que busques el jugador óptimo de hace 3 semanas, además porque habría que almacenar en la base de datos información de la que ya carecemos. Por tanto, lo que no está claro es el nombre de la función que da lugar a error.

javigp2002 replied:
Debido a que para los tests, tengo que crear un calendario fictició. Aunque, viendolo así, puedo crear ese calendario con día de hoy también 💯

javigp2002 replied:
Efectivamente hubo un problema al principio cunado se estableció el `Calendario` puesto que se podia introducir una fecha para conocer el óptimo pero eso no es óptimo 😄. Resuelto en:;  [6dd5895](https://github.com/javigp2002/LazyFood/pull/84/commits/6dd589550b2fb234075848d41dad66de5d1861c5), [1388bff](https://github.com/javigp2002/LazyFood/pull/84/commits/1388bff56ac16472fcc32ad4fdb79e89dd274582)

JJ reviewed (changes requested):
Lo siento, pero siendo este el último objetivo me resisto a darlo por aprobado si tienes ciertos antipatrones en el código. Tienes todavía unos días para modificarlo, no creo que te cueste trabajo. Si no estás de acuerdo con lo que digo, podemos hablarlo

javigp2002 committed:
:zap: Improve: SRP en create bd, closes https://github.com/javigp2002/LazyFood/pull/84#discussion_r1448850643

Se ha realizado un cambio para poder contemplar mejor lo que devolvería un create de la bd

javigp2002 committed:
:rocket: Modify: escoger el jugador óptimo según fecha, clsoes #93

Debido a https://github.com/javigp2002/LazyFood/pull/84#discussion_r1448856118 está claro que no es necesario ir pasando la fecha por cada vez que queramos saber el óptimo de un jugador debido a que la bbdd no almacena más de 5 partidos y no podemos conocer un óptimo actual, ni tendría sentido

javigp2002 committed:
:white_check_mark: Update: tests, closes #94

Con el cambio debido al #93, debemos actualizar los tests

javigp2002 commented:
@JJ Listo para revisión 🚀

JJ commented on src/api_controller.ts:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    async getOptimo(nombreEquipo: string): Promise<Jugador|null> {
+        const res = await this.db.getOptimo(nombreEquipo);
```
Sigues el mismo patrón const x, return x. ¿Crees que la creación de esa variable es necesaria?

javigp2002 replied:
Cierto, es que antes hacía más cosas y se habrá quedado...  [c948287](https://github.com/javigp2002/LazyFood/pull/84/commits/c94828775bfcc0f04ec7b3fb95cef56817197d39)

JJ reviewed (changes requested):
Lo siento, Javi; has corregido muchas cosas, pero el tema de los códigos de estado, que es importante para diseñar adecuadamente un API, no parece que siga las convenciones de REST. Por su importancia [he añadido un item al respecto al fichero de errores frecuentes](https://github.com/JJ/IV-/blob/master/errores/errores-objetivo-9.md).
Siento entretenerte más con esto, pero ya te falta muy poquito. Ánimo.

javigp2002 committed:
:bug: Creación innecesaria de variable, closes https://github.com/javigp2002/LazyFood/pull/84#discussion_r1450367713

Se estaba creando una variable innecesaria a las respuestas del api_controller

javigp2002 committed:
Fix status POST, closes #https://github.com/javigp2002/LazyFood/pull/84#discussion_r1450366995

El post cuando se crea debe devolver un estado diferente si es creado a si estaba ya, pero no un error si ya estaba creado en la bd. Por otro lado, se responderá en el 201 con la data recibida

javigp2002 commented on src/api_controller.ts:
```diff
@@ -0,0 +1,51 @@
+
+    async postJugador(body: any) {
+        const is_created = await this.db.createJugador(body);
+        return is_created ? 200 : 400;
```
Resuelto en [38833b2](https://github.com/javigp2002/LazyFood/pull/84/commits/38833b2d740ad9d7026e73654015688cf29409df)

javigp2002 commented:
@JJ gracias por su tiempo, lo primero. No se disculpe las cosas o están bien o no se aprueba de toda la vida. Listo para revision!

JJ reviewed (approved):
:clap:
